libethereum/Client.cpp

```diff
@@ -858,7 +858,30 @@ pair<h256, Address> Client::submitTransaction(TransactionSkeleton const& _t, Sec
         ts.gas = defaultTransactionGas;
 
     Transaction t(ts, _secret);
-    m_tq.import(t.rlp());
+
+    // Use the Executive to perform basic validation of the transaction
+    // (e.g. transaction signature, account balance) using the state of
+    // the latest block in the client's blockchain. This can throw but
+    // we'll catch the exception at the RPC level.
+    Block currentBlock = block(bc().currentHash());
+    Executive e(currentBlock, bc());
+    e.initialize(t);
+    ImportResult res = m_tq.import(t.rlp());
+    switch (res)
+    {
+        case ImportResult::Success:
+            break;
+        case ImportResult::ZeroSignature:
+            BOOST_THROW_EXCEPTION(ZeroSignatureTransaction());
+        case ImportResult::OverbidGasPrice:
+            BOOST_THROW_EXCEPTION(GasPriceTooLow());
+        case ImportResult::AlreadyKnown:
+            BOOST_THROW_EXCEPTION(PendingTransactionAlreadyExists());
+        case ImportResult::AlreadyInChain:
+            BOOST_THROW_EXCEPTION(TransactionAlreadyInChain());
+        default:
+            BOOST_THROW_EXCEPTION(UnknownError());
```

Comment:
I'd prefer this to be a little more specific, like `UnknownTransactionValidationError`